In [3]:
import pandas as pd
import numpy as np

In [4]:
print("working now")

working now


In [5]:
df = pd.read_csv("../docs/raw/BigBasket_Products.csv")
df.head()

,index,product,category,sub_category,brand,sale_price,market_price,type,rating,description
0,1,Garlic Oil - Vegetarian Capsule 500 mg,Beauty & Hygiene,Hair Care,Sri Sri Ayurveda,220.0,220.0,Hair Oil & Serum,4.1,This Product contains Garlic Oil that is known...
1,2,Water Bottle - Orange,"Kitchen, Garden & Pets",Storage & Accessories,Mastercook,180.0,180.0,Water & Fridge Bottles,2.3,"Each product is microwave safe (without lid), ..."
2,3,"Brass Angle Deep - Plain, No.2",Cleaning & Household,Pooja Needs,Trm,119.0,250.0,Lamp & Lamp Oil,3.4,"A perfect gift for all occasions, be it your m..."
3,4,Cereal Flip Lid Container/Storage Jar - Assort...,Cleaning & Household,Bins & Bathroom Ware,Nakoda,149.0,176.0,"Laundry, Storage Baskets",3.7,Multipurpose container with an attractive desi...
4,5,Creme Soft Soap - For Hands & Body,Beauty & Hygiene,Bath & Hand Wash,Nivea,162.0,162.0,Bathing Bars & Soaps,4.4,Nivea Creme Soft Soap gives your skin the best...


In [7]:
df.shape

(27555, 10)

In [8]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 27555 entries, 0 to 27554
Data columns (total 10 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   index         27555 non-null  int64  
 1   product       27554 non-null  str    
 2   category      27555 non-null  str    
 3   sub_category  27555 non-null  str    
 4   brand         27554 non-null  str    
 5   sale_price    27555 non-null  float64
 6   market_price  27555 non-null  float64
 7   type          27555 non-null  str    
 8   rating        18929 non-null  float64
 9   description   27440 non-null  str    
dtypes: float64(3), int64(1), str(6)
memory usage: 2.1 MB


In [9]:
df.describe()

,index,sale_price,market_price,rating
count,27555.00000,27555.000000,27555.000000,18929.000000
mean,13778.00000,322.514808,382.056664,3.943410
std,7954.58767,486.263116,581.730717,0.739063
min,1.00000,2.450000,3.000000,1.000000
25%,6889.50000,95.000000,100.000000,3.700000
50%,13778.00000,190.000000,220.000000,4.100000
75%,20666.50000,359.000000,425.000000,4.300000
max,27555.00000,12500.000000,12500.000000,5.000000


In [10]:
df.isnull().sum()

index              0
product            1
category           0
sub_category       0
brand              1
sale_price         0
market_price       0
type               0
rating          8626
description      115
dtype: int64

In [11]:
df.duplicated().sum()

np.int64(0)

In [12]:
df['category'].value_counts()

category
Beauty & Hygiene            7867
Gourmet & World Food        4690
Kitchen, Garden & Pets      3580
Snacks & Branded Foods      2814
Foodgrains, Oil & Masala    2676
Cleaning & Household        2675
Beverages                    885
Bakery, Cakes & Dairy        851
Baby Care                    610
Fruits & Vegetables          557
Eggs, Meat & Fish            350
Name: count, dtype: int64

In [13]:
df['sub_category'].value_counts()

sub_category
Skin Care                   2294
Health & Medicine           1133
Hair Care                   1028
Storage & Accessories       1015
Fragrances & Deos           1000
                            ... 
Mutton & Lamb                 20
Water                         17
Flower Bouquets, Bunches       8
Pork & Other Meats             5
Marinades                      1
Name: count, Length: 90, dtype: int64

In [14]:
df['brand'].value_counts().head(10)

brand
Fresho              638
bb Royal            539
BB Home             428
DP                  250
Fresho Signature    171
bb Combo            168
Amul                153
INATUR              146
Himalaya            141
Dabur               138
Name: count, dtype: int64

In [15]:
df[['sale_price', 'market_price']].describe()

,sale_price,market_price
count,27555.000000,27555.000000
mean,322.514808,382.056664
std,486.263116,581.730717
min,2.450000,3.000000
25%,95.000000,100.000000
50%,190.000000,220.000000
75%,359.000000,425.000000
max,12500.000000,12500.000000


In [20]:
df.nunique()

index               27555
product             23540
category               11
sub_category           90
brand                2313
sale_price           3256
market_price         1348
type                  426
rating                 40
description         21944
discount_percent     2662
dtype: int64

In [18]:
df['discount_percent'] = round(((df['market_price'] - df['sale_price']) / df['market_price']) * 100,2)

df.head()

,index,product,category,sub_category,brand,sale_price,market_price,type,rating,description,discount_percent
0,1,Garlic Oil - Vegetarian Capsule 500 mg,Beauty & Hygiene,Hair Care,Sri Sri Ayurveda,220.0,220.0,Hair Oil & Serum,4.1,This Product contains Garlic Oil that is known...,0.00
1,2,Water Bottle - Orange,"Kitchen, Garden & Pets",Storage & Accessories,Mastercook,180.0,180.0,Water & Fridge Bottles,2.3,"Each product is microwave safe (without lid), ...",0.00
2,3,"Brass Angle Deep - Plain, No.2",Cleaning & Household,Pooja Needs,Trm,119.0,250.0,Lamp & Lamp Oil,3.4,"A perfect gift for all occasions, be it your m...",52.40
3,4,Cereal Flip Lid Container/Storage Jar - Assort...,Cleaning & Household,Bins & Bathroom Ware,Nakoda,149.0,176.0,"Laundry, Storage Baskets",3.7,Multipurpose container with an attractive desi...,15.34
4,5,Creme Soft Soap - For Hands & Body,Beauty & Hygiene,Bath & Hand Wash,Nivea,162.0,162.0,Bathing Bars & Soaps,4.4,Nivea Creme Soft Soap gives your skin the best...,0.00


In [19]:
df['discount_percent'].describe()

count    27555.000000
mean        11.824860
std         14.624493
min          0.000000
25%          0.000000
50%          5.000000
75%         20.000000
max         83.670000
Name: discount_percent, dtype: float64

### Key Observations

## Missing Values

1. rating — 8,626 nulls (~31.3%) — This is the most critical gap. Over 1/3rd of products have no rating, which could heavily impact any recommendation or ranking model.
2. description — 115 nulls (~0.4%) — Minor but worth imputing or filtering for NLP/embedding-based tasks.

## No Duplicates

## Pricing Insights

1. High standard deviation relative to mean → significant price skew / long-tail distribution (many cheap items, a few very expensive ones).
2. The gap between sale and market price suggests discounting is common.

## Engineered Feature

1. Mean discount: ~11.8%
2. Median discount: 5% (50th percentile)
3. Max discount: 83.67% — some products are heavily discounted
4. 25th percentile = 0% → at least a quarter of products have no discount at all

## Ratings Distribution

1. Rated products only (18,929 records)
2. Mean: 3.94 out of 5
3. Std Dev: 0.74 — fairly concentrated
4. 25th–75th percentile: 3.7 to 4.3 — most products cluster in a "good" range
5. Only 40 unique rating values (likely increments of 0.1)

## Category Dominance

Beauty & Hygiene dominates at nearly 3× the second category — expect class imbalance if building category classifiers.


## Recommendations

1. Handle rating nulls as most of the customers didnt review the product they purchased.
2. description nulls to fill with empty string or a placeholder like "No description available"
3. Category imbalance as Beauty & Hygiene is ~28.5% of the data; using stratified sampling when splitting can help us give some accurate results.
